### ISSUE TITLE
Lint which reports on initialization of late final field or variable from a method or function

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#2444]
**Describe the rule you'd like to see implemented**

Disclaimer up front: It is possible that this lint would have so many false positives as to really diminish its usefulness. I'd love to discuss.

In order to encourage safe use of `late final` fields and variables (and perhaps to _discourage_ declaring them at all), I'd like a lint which reports when such a field or variable is initialized in a place which can be executed more than once. The idea is that such code may be called twice, which would throw an error at runtime. For example:

```dart
class C {
  late final int a;
  initializeA() => a = 7; // LINT
}
```

The places where it would be _safe_ to initialize a `late final` field include:

* a constructor, which is only executed once for any given object
* the field's initializer

Examples of places where it would be _unsafe_ to initialize a `late final` field include:

* a public method on the class
* a function outside the class

Maybe we say that private methods are safe, since the owner of the field owns the private method, and can be assumed responsible.

This lint rule would basically restrict all `late final` fields to the following two uses:

1. I want a `final` field which is initialized lazily, only when it is read. In order to initialize the field, I may need to access `this`. For example:

   ```dart
   class C {
     final List<Widget> widgets;
     late final List<Widget> widgetsFilteredByFoo = () {
       return [
         for (var widget in this.widgets)
           if (widget.someExpensiveFilterCalculation())
             widget
         ];
       }();
   }
   ```

2. I want a `final` field which is initialized in a constructor body, rather than a constructor initializer, in order to use complex logic, or access `this`.
